osquery/tables/system/windows/shortcut_files.cpp

```diff
@@ -0,0 +1,214 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <osquery/core/core.h>
+#include <osquery/core/tables.h>
+#include <osquery/filesystem/filesystem.h>
+#include <osquery/tables/system/windows/registry.h>
+#include <osquery/utils/conversions/join.h>
+
+#include <osquery/logger/logger.h>
+#include <osquery/utils/windows/shelllnk.h>
+
+#include <boost/filesystem.hpp>
+
+#include <sstream>
+#include <string>
+
+namespace osquery {
+namespace tables {
+
+QueryData genShortcutFiles(QueryContext& context) {
+  QueryData results;
+  auto paths = context.constraints["path"].getAll(EQUALS);
+  // Expand contraints
+  context.expandConstraints(
+      "path",
+      LIKE,
+      paths,
+      ([&](const std::string& pattern, std::set<std::string>& out) {
+        std::vector<std::string> patterns;
+        auto status =
+            resolveFilePattern(pattern, patterns, GLOB_ALL | GLOB_NO_CANON);
+        if (status.ok()) {
+          for (const auto& resolved : patterns) {
+            out.insert(resolved);
+          }
+        }
+        return status;
+      }));
+
+  boost::system::error_code ec;
+  for (const auto& lnk : paths) {
+    boost::filesystem::path path = lnk;
+    if (!boost::filesystem::is_regular_file(path, ec)) {
+      continue;
+    }
+    std::string lnk_content;
+    if (!readFile(path, lnk_content, 20).ok()) {
+      LOG(WARNING) << "Failed to read shortcut file: " << lnk;
+    }
+    std::stringstream check_ss;
+    for (const auto& hex_char : lnk_content) {
+      std::stringstream value;
+      value << std::hex << std::uppercase << (int)(unsigned char)(hex_char);
+      // Add additional 0 if single hex value is 0-F
+      if (value.str().size() == 1) {
+        check_ss << "0";
+      }
+      check_ss << value.str();
+    }
+    std::string header_sig = check_ss.str();
+
+    // Check for shortcut header size (0x0000004c) and GUID
+    // (00021401-0000-0000-c000-000000000046)
+    if (header_sig != "4C0000000114020000000000C000000000000046") {
+      continue;
+    }
+
+    // Read the whole shortcut file
+    lnk_content = "";
+    if (!readFile(path, lnk_content).ok()) {
+      LOG(WARNING) << "Failed to read shortcut file: " << lnk;
+    }
+    std::stringstream ss;
+    for (const auto& hex_char : lnk_content) {
+      std::stringstream value;
+      value << std::hex << std::uppercase << (int)(unsigned char)(hex_char);
+      // Add additional 0 if single hex value is 0-F
+      if (value.str().size() == 1) {
+        ss << "0";
+      }
+      ss << value.str();
+    }
```

Comment:
so before reading the whole shortcut file, i added a check to read the first 20 byte to check for the expected shortcut header size and the unique shortcut GUID.
So the first for loop is for the first readFile for checking the header. If its a valid shortcut header, the i call readFile again and read the whole file. 
I do another for loop which is the same as the loop for the header check

Reply:
Maybe that's a common idiom? but it seems like it should be cleaner to read once, check if the first 20 bytes match, and then continue or early return

Reply:
So i added check to read only the first 20 bytes to avoid a user accidentally read a very large file if they use wildcards  
But if thats not a concern i can remove it and merge it into one read

Reply:
Yeah... This probably makes more sense as is.